test/utilities.jl

```diff
@@ -0,0 +1,128 @@
+"""
+assert_allclose(actual, reference; [rtol], [atol], [err_msg], [error_location_fmt])
+
+Raises an assertion exception when any entries in `actual` and `reference` are not equal up to the
+desired tolerance. In other words, all entries of `actual` and `reference` must satisfy:
+`abs(actual - reference) <= rtol * abs(reference) + atol`
+
+This is inspired by numpy's assert_allclose function. However, we've changed some default values.
+Additionally, in cases where actual and reference don't have equal values but do have the same 
+shape, the locations of the max errors are also printed.
+
+# Keywords
+- `rtol`: The desired relative tolerance
+- `atol`: The desired absolute tolerance
+- `err_msg`: An optional error message that can be printed when the arrays are not close
+- `error_location_fmt`: An optional keyword used to specify a function that creates a string 
+  describing the location in the arrays where values are unequal, given a `CartesianIndex` object.
+  This should not be specified in cases where actual and reference have different shapes.
+"""
+function assert_allclose(actual, reference; rtol = 1e-7, atol = 0.0, err_msg = nothing,
+                         error_location_fmt = nothing)
+    # make the initial check fast!
+    if !any( abs.(actual .- reference) .> (rtol .* abs.(reference) .+ atol))
+        return true
+    end
+
+    common_layout = (size(actual) == size(reference)) && (ndims(actual) == ndims(reference))
+    if isnothing(error_location_fmt)
+        error_location_fmt = (arg) -> string(arg.I)
+    elseif !common_layout
+        throw("error_location_fmt must be nothing b/c actual and reference have different shapes")
+    end
+
+    # now let's provide a detailed error message (this can take longer):
+
+    # first, let's determine what the max error is (and possibly where it happended)
+    max_rel_error_magnitude = -1.0
+    max_rel_error_magnitude_loc = nothing
+    max_abs_diff = 0.0
+    max_abs_diff_loc = nothing
+    function func(actual_val, ref_val; index = nothing)
+        cur_abs_diff = abs(actual_val - ref_val)
+        if max_abs_diff < cur_abs_diff
+            max_abs_diff = cur_abs_diff
+            max_abs_diff_loc = index
+        end
+
+        if ref_val != 0.0
+            cur_rel_error_mag = cur_abs_diff / abs(ref_val)
+            if max_rel_error_magnitude < cur_rel_error_mag
+                max_rel_error_magnitude = cur_rel_error_mag
+                max_rel_error_magnitude_loc = index
+            end
+        end
+    end
+
+    if common_layout
+        for index in CartesianIndices(size(actual))
+            func(actual[index], reference[index]; index = index)
+        end
+    else
+        func.(actual, reference)
+    end
+
```

Comment:
Does an implementation like this make sense?  Are you getting anything by avoiding broadcasting in the `common_layout` case?

```julia
diff = abs.(actual_val .- ref_val)
max_abs_ind = argmax(diff)
max_abs_val = diff[max_abs_diff]

relative_diff = diff ./ abs.(ref_val)
max_rel_ind = argmax.(relative_dif)
max_rel_val = max_rel_diff[max_rel_ind]
```